File: lana/src/main/scala/com/financialforce/lana/commands/ShowLogFile.scala
/*
 * Copyright (c) 2020 FinancialForce.com, inc. All rights reserved.
 */
package com.financialforce.lana.commands

import com.financialforce.lana.display.QuickPickWorkspace
import com.financialforce.lana.runtime.vscode.{Uri, window}
import com.financialforce.lana.{Context, Main}
import io.scalajs.nodejs.fs.Fs
import io.scalajs.nodejs.path.Path

import scala.concurrent.Future
import scala.scalajs.concurrent.JSExecutionContext.Implicits.queue
import scala.scalajs.js

object ShowLogFile extends LogView {
  def apply(context: Context): Unit = {
    Command("showLogFile", (uri: Uri) => command(context, uri)).register(context)
    context.display.output(s"Registered command '${Main.appName}: Show Log'")
  }

  private def command(context: Context, uri: Uri): Future[Unit] = {
    val filePath = if (js.isUndefined(uri)) {
      window.activeTextEditor.map(_.document.uri.fsPath).toOption
    } else {
      Some(uri.fsPath)
    }

    filePath.map(filePath => {
      val cmd =
        QuickPickWorkspace.pickOrReturn(context) flatMap { wsPath =>
          val name = Path.parse(filePath).name.getOrElse("Unknown")
          val fileContents = Fs.readFileSync(filePath, "utf-8")
          createView(wsPath, context, name, filePath, fileContents)
          Future.successful(())
        }
      cmd.failed foreach { ex =>
        context.display.showErrorMessage(ex.getMessage)
      }
      cmd
    }).getOrElse {
      context.display.showErrorMessage("No file selected to display log analysis")
      Future.successful(())
    }
  }
}

File: src/commands/ShowLogFile.ts
/*
 * Copyright (c) 2020 FinancialForce.com, inc. All rights reserved.
 */

import { Uri, window } from "vscode";
import { Context } from "../Context";
import { QuickPickWorkspace } from "../display/QuickPickWorkspace";
import * as path from "path";
import { promises as fs } from "fs";
import { LogView } from "./LogView";
import { Command } from "./Command";
import { appName } from "../Main";

export class ShowLogFile {
  static apply(context: Context): void {
    new Command("showLogFile", (uri: Uri) =>
      ShowLogFile.safeCommand(context, uri)
    ).register(context);
    context.display.output(`Registered command '${appName}: Show Log'`);
  }

  private static async safeCommand(context: Context, uri: Uri): Promise<void> {
    try {
      return ShowLogFile.command(context, uri);
    } catch (err) {
      context.display.showErrorMessage(`Error showing logfile: ${err.message}`);
      return Promise.resolve();
    }
  }

  private static async command(context: Context, uri: Uri): Promise<void> {
    const filePath = uri?.fsPath || window?.activeTextEditor?.document.fileName;

    if (filePath) {
      const name = path.parse(filePath).name;
      const ws = await QuickPickWorkspace.pickOrReturn(context);

      const [view, fileContent] = await Promise.all([
        LogView.createView(ws, context, name),
        fs.readFile(filePath, "utf-8"),
      ]);

      LogView.appendView(view, context, name, filePath, fileContent);
    } else {
      context.display.showErrorMessage(
        "No file selected to display log analysis"
      );
      throw new Error("No file selected to display log analysis");
    }
  }
}
